### PR TITLE
Feature/marineglider

### PR DESCRIPTION
**Pull request by lliu1220**
This is a draft PR.

This PR is to address @srherbener 's request to merge temperature and salinity into one converter per closed PR https://github.com/JCSDA-internal/ioda-converters/pull/341#issuecomment-721442057.

This PR will be addressed under soca-science related to how to ingest AOML hurricane glider data into soca-science https://github.com/JCSDA-internal/soca-science/issues/83

To test the converter on Orion:

1. Download branch feature/marineglider:**
    git clone https://github.com/JCSDA-internal/ioda-converters.git
    git checkout feature/marineglider
    
2 module use -a /work/noaa/marine/lliu/modulefiles/
    module load anaconda

3. cd src/marine
    python glider2ioda.py -i test/testinput/SG547_0001_dn_AOML.nc -o test/testoutput/test_glider.nc -d 2016080412
    (this reads in testing AOML glider data SG547_0001_dn_AOML.nc and write both temperature and salinity into output ioda file test_glider.nc )

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -84,6 +84,10 @@ Usage: <converter.py> -i INPUT_FILE(S) -o OUTPUT_FILE -d YYYYMMDDHH
   * `hgodas_adt2ioda.py`
   * `hgodas_insitu2ioda.py`
   * `hgodas_sst2ioda.py`  
+
+* AOML hurricane glider - preprocessed, QCd S/T from AOML hurricane glider for 2020 hurricane track analysis project
+  * glider2ioda.py`
```
Comment (travissluka) [inline README.md:89]:
```suggestion
  * `glider2ioda.py`
```

File: src/marine/ioda_conv_ncio.py [hunk 2]
```diff
@@ -0,0 +1,472 @@
+#!/usr/bin/env python
```
Comment (travissluka) [inline src/marine/ioda_conv_ncio.py:1]: why is this file here? delete

File: src/marine/orddicts.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+from collections import OrderedDict, Callable
```
Comment (travissluka) [inline src/marine/orddicts.py:1]: delete this file

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -18,6 +18,7 @@ list( APPEND test_input
   testinput/smap_sss_rss.nc
   testinput/hgodas_insitu.nc
   testinput/hgodas_sst.nc
+  testinput/SG547_0001_dn_AOML.nc
```
Comment (travissluka) [inline test/CMakeLists.txt:21]: can you rename this to something more meaningful? i have no idea this is glider data given the filename. Ideally have input and output filenames the same.

**Review by travissluka — CHANGES_REQUESTED** [review]
fix conflicts with merging develop.
Fix coding norms, make sure you are running `ctest` within the `<build_dir>/ioda-converters` directory, not `<build_dir>/ioda_converters/test` directory (otherwise those coding norm tests don't run)

File: README.md (hunk 1 above)
Comment (lliu1220) [reply README.md:89]: Three things:

1, if I run ctest under {build_dir}/ioda-converters/test, all ctests including glider but one (ncep_bufr) failed:

 5/24 Test  #5: test_iodaconv_godae_prof ..........   Passed    0.50 sec
      Start  6: test_iodaconv_godae_ship
 6/24 Test  #6: test_iodaconv_godae_ship ..........   Passed    0.49 sec
      Start  7: test_iodaconv_godae_trak
 7/24 Test  #7: test_iodaconv_godae_trak ..........   Passed    0.43 sec
      Start  8: test_iodaconv_hgodas_insitu
 8/24 Test  #8: test_iodaconv_hgodas_insitu .......   Passed    0.37 sec
      Start  9: test_iodaconv_glider
 **9/24 Test  #9: test_iodaconv_glider ..............   Passed    0.39 sec
      Start 10: test_iodaconv_hgodas_sst**
10/24 Test #10: test_iodaconv_hgodas_sst ..........   Passed    0.40 sec
      Start 11: test_iodaconv_argoclim
11/24 Test #11: test_iodaconv_argoclim ............   Passed    0.40 sec
      Start 12: test_iodaconv_cryosat2
12/24 Test #12: test_iodaconv_cryosat2 ............   Passed    0.59 sec
      Start 13: test_iodaconv_viirs_jpss1_oc_l2
96% tests passed, 1 tests failed out of 24

Total Test time (real) =  11.05 sec

The following tests FAILED:
         **15 - test_iodaconv_ncep_bufr (Failed)**

2. however, if i ran ctest under {build_dir}/ioda-converters, then 8 failed out of 32 (coding norm error), but **glider still passed,** so my question is: so seems coding norm error does not apply to glider but the other 8 of ctests, is this something that I should work on since glider passed?

14/31 Test #14: test_iodaconv_godae_trak .............   Passed    0.49 sec
      Start 15: test_iodaconv_hgodas_insitu
15/31 Test #15: test_iodaconv_hgodas_insitu ..........   Passed    0.45 sec
      **Start 16: test_iodaconv_glider
16/31 Test #16: test_iodaconv_glider .................   Passed    0.46 sec**
      Start 17: test_iodaconv_hgodas_sst
17/31 Test #17: test_iodaconv_hgodas_sst .............   Passed    0.43 sec

74% tests passed, 8 tests failed out of 31

Total Test time (real) =  15.08 sec

The following tests FAILED:
          1 - iodaconv_chem_coding_norms (Failed)
          2 - iodaconv_gsi-ncdiag_coding_norms (Failed)
          3 - iodaconv_lib-python_coding_norms (Failed)
          4 - iodaconv_marine_coding_norms (Failed)
          5 - iodaconv_ncep_coding_norms (Failed)
          6 - iodaconv_ssec_coding_norms (Failed)
          7 - iodaconv_wrfda-ncdiag_coding_norms (Failed)
         22 - test_iodaconv_ncep_bufr (Failed)

3. If you still need me to work on coding norm (for the other failed tests), can you provide instruction? how am I supposed to test coding norm conflict and how to fix, cause I have no tool of testing nor fixing .

4. Last, I removed files you asked and renamed glider input file.

**Comment by travissluka** [comment]
@lliu1220 run
```
git fetch --all
git merge origin/develop
``` 
and fix the resulting conflicts before doing anything else. You haven't merged in the the latest develop and you are a month behind. You'll see that there have been changes to the tests

**Comment by lliu1220** [comment]
git clone https://github.com/JCSDA-internal/ioda-converters.git
git fetch --all
git merge origin/develop

ecbuild --prefix=/work/noaa/da/lliu/tools ../
make
make install

ctest:

70% tests passed, 10 tests failed out of 33

Label Time Summary:
iodaconv    =   8.89 sec*proc (32 tests)
script      =   8.89 sec*proc (32 tests)

Total Test time (real) =   8.92 sec

The following tests FAILED:
          1 - iodaconv_chem_coding_norms (Failed)
          2 - iodaconv_gsi-ncdiag_coding_norms (Failed)
          3 - iodaconv_land_coding_norms (Failed)
          4 - iodaconv_lib-python_coding_norms (Failed)
          5 - iodaconv_marine_coding_norms (Failed)
          6 - iodaconv_metar_coding_norms (Failed)
          7 - iodaconv_ncep_coding_norms (Failed)
          8 - iodaconv_ssec_coding_norms (Failed)
          9 - iodaconv_wrfda-ncdiag_coding_norms (Failed)
         24 - test_iodaconv_ncep_bufr (Failed)

**Comment by lliu1220** [comment]
Yay!--

      Start  4: iodaconv_marine_coding_norms
 4/31 Test  #4: iodaconv_marine_coding_norms .........   Passed    0.37 sec
      Start 16: test_iodaconv_glider
16/31 Test #16: test_iodaconv_glider .................   Passed    0.29 sec

**Comment by lliu1220** [comment]
can you be more explicit what is the conflict? this does not state what the conflict is nor how to resolve it, so confusing.
image

**Comment by travissluka** [comment]
@lliu1220 the code in your branch is between `<<<<<<< feature/marineglider` and `=======` lines, the code in develop is between `=======` and >>>>>>> develop`. as you can see they are complete different. you need to manually edit the file to merge your changes into what is shown by the develop branch

**Comment by lliu1220** [comment]
I'm closing this PR since this branch was month old,
new branch feature/gliderAOML.
new PR, https://github.com/JCSDA-internal/ioda-converters/pull/362

**Comment by travissluka** [comment]
@lliu1220, i'm not sure how you were trying to fix the merge conflicts, but for future reference the following should have worked fine:

with your current branch checked out:
```
git fetch --all
git merge origin/develop
```
git will complain with the following
```
Auto-merging test/CMakeLists.txt
CONFLICT (content): Merge conflict in test/CMakeLists.txt
Auto-merging src/marine/CMakeLists.txt
Auto-merging README.md
Automatic merge failed; fix conflicts and then commit the result.
```
after you fix the merge conflicts in `test/CMakeLists.txt` run
```
git commit
```
done.

**Comment by lliu1220** [comment]
@travissluka 

the problem is this old branch was too behind, I could spend tooooo much time trying to fix every single one of the conflicts, alternatively, I created new branch and done in 5 minutes.

Humor me we both spent too much time on this, can we move to this PR https://github.com/JCSDA-internal/ioda-converters/pull/362, and get done with it in just 5 minutes.

Thank you for all the tips.
